marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -229,40 +438,34 @@ evalObservation env state obs = let
 
 -- | Pick the first account with money in it
 refundOne :: Map AccountId Money -> Maybe ((Party, Money), Map AccountId Money)
-refundOne accounts = do
-    ((accId, money), rest) <- Map.minViewWithKey accounts
-    if getLovelace money > 0
-    then return ((accountOwner accId, money), rest)
-    else refundOne rest
-
-
-data Payment = Payment Party Money
-  deriving (Eq,Ord,Show,Read)
-
-data ReduceEffect = ReduceWithPayment Payment
-                  | ReduceNoPayment
-  deriving (Eq,Ord,Show,Read)
+refundOne accounts = case Map.toList accounts of
+    [] -> Nothing
+    (accId, balance) : rest ->
+        if balance > 0
+        then Just ((accountOwner accId, balance), Map.fromList rest)
+        else refundOne (Map.fromList rest)
 
 
 -- | Obtains the amount of money available an account
 moneyInAccount :: AccountId -> Map AccountId Money -> Money
-moneyInAccount = Map.findWithDefault (Lovelace 0)
+moneyInAccount accId accounts = case Map.lookup accId accounts of
+    Just x  -> x
+    Nothing -> Ada.lovelaceOf 0
 
 
 -- | Sets the amount of money available in an account
 updateMoneyInAccount :: AccountId -> Money -> Map AccountId Money -> Map AccountId Money
 updateMoneyInAccount accId money =
-    if getLovelace money <= 0 then Map.delete accId else Map.insert accId money
+    if Ada.getLovelace money <= 0 then Map.delete accId else Map.insert accId money
```

Comment:
Why not just do comparisons on `Ada`? You're probably noticing a theme here, I'll stop talking about it. But generally, I think you shouldn't need to be converting to/from integers all the time.

Reply:
I'm about to change this code as part of multicurrency support in Marlowe, so it's not that important right now.